From d7dc7ccebec804c479e1392f5d49c825d5ecc0c3 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Sat, 20 Mar 2021 18:12:12 -0500
Subject: [PATCH] Increase script numbers to 63-bit + sign bit (#71)

Closes #80.

Rebase provenance: this historical change was adapted onto the pinned
Bitcoin ABC base. Source diff is the recorded Lotus V1 replay commit below.

Historical-source-commit: a0d5d372f202384bf31d8e143f9500ba17af7325
Recorded-replay-commit: dcd7e365e86c21b28817cc52c839ccc4c261319e

diff --git a/src/script/script.h b/src/script/script.h
index 2857becee2..1310db6b0e 100644
--- a/src/script/script.h
+++ b/src/script/script.h
@@ -226,13 +226,9 @@ struct scriptnum_encoding_error : public std::runtime_error {
 
 class CScriptNum {
     /**
-     * Numeric opcodes (OP_1ADD, etc) are restricted to operating on 4-byte
-     * integers. The semantics are subtle, though: operands must be in the range
-     * [-2^31 +1...2^31 -1], but results may overflow (and are valid as long as
-     * they are not used in a subsequent numeric operation). CScriptNum enforces
-     * those semantics by storing results as an int64 and allowing out-of-range
-     * values to be returned as a vector of bytes but throwing an exception if
-     * arithmetic is done or the result is interpreted as an integer.
+     * Numeric opcodes (OP_1ADD, etc) are restricted to operating on 8-byte
+     * integers. Both operands and results must be in the range
+     * [-2^63 +1...2^63 -1], throwing an exception otherwise.
      */
 public:
     explicit CScriptNum(const int64_t &n) { m_value = n; }
@@ -302,6 +298,9 @@ public:
     }
 
     inline CScriptNum operator/(const int64_t &rhs) const {
+        // Sanity check for invalid script num; should never occur
+        assert(m_value != std::numeric_limits<int64_t>::min());
+        assert(rhs != 0);
         return CScriptNum(m_value / rhs);
     }
     inline CScriptNum operator/(const CScriptNum &rhs) const {
@@ -309,6 +308,9 @@ public:
     }
 
     inline CScriptNum operator%(const int64_t &rhs) const {
+        // Sanity check for invalid script num; should never occur
+        assert(m_value != std::numeric_limits<int64_t>::min());
+        assert(rhs != 0);
         return CScriptNum(m_value % rhs);
     }
     inline CScriptNum operator%(const CScriptNum &rhs) const {
diff --git a/src/test/data/script_tests.json b/src/test/data/script_tests.json
index a616939c00..1bf2e0e517 100644
--- a/src/test/data/script_tests.json
+++ b/src/test/data/script_tests.json
@@ -955,7 +955,7 @@
 ["RAWLEFTBITSHIFT"],
 ["", "RAWLEFTBITSHIFT 1", "", "INVALID_STACK_OPERATION", "Empty stack"],
 ["0x02 0x1234", "RAWLEFTBITSHIFT 1", "", "INVALID_STACK_OPERATION", "Not enough items on stack"],
-["0x02 0x1234", "'abcdefghijklm' RAWLEFTBITSHIFT 1", "", "UNKNOWN_ERROR", "Bad size"],
+["0x02 0x1234", "'abcdefghijklm' RAWLEFTBITSHIFT 1", "", "INTEGER_OVERFLOW", "Bad size"],
 ["0x02 0x1234", " 1 RAWLEFTBITSHIFT 0x02 0x2468 EQUAL", "", "OK", "simple left shift"],
 ["0x02 0x1234", " 2 RAWLEFTBITSHIFT 0x02 0x48d0 EQUAL", "", "OK", "simple left shift"],
 ["0x02 0x1234", " 3 RAWLEFTBITSHIFT 0x02 0x91a0 EQUAL", "", "OK", "simple left shift"],
@@ -1348,10 +1348,6 @@
 ["'abcdefghi'", "0NOTEQUAL", "", "INTEGER_OVERFLOW", "0NOTEQUAL is an arithmetic operand"],
 
 ["ADD"],
-["2147483648 0 ADD", "2147483648 EQUAL", "", "OK", "int64 operands are fine"],
-["-2147483648 0 ADD", "-2147483648 EQUAL", "", "OK", "int64 operands are fine"],
-["2147483647 DUP ADD", "4294967294 NUMEQUAL", "", "DISABLED_OPCODE", "NUMEQUAL is disabled"],
-["'abcdefabc' NOT", "0 EQUAL", "", "INTEGER_OVERFLOW", "NOT is an arithmetic operand"],
 ["0x09 0x000000000000000001 0", "ADD 1", "", "INTEGER_OVERFLOW", "arithmetic operands must be in range [-2^63 + 1...2^63 - 1]"],
 ["0x09 0x000000000000000081 0", "ADD 1", "", "INTEGER_OVERFLOW", "arithmetic operands must be in range [-2^63 + 1...2^63 - 1]"],
 ["9223372036854775807", "DUP ADD 1", "", "INTEGER_OVERFLOW", "ADD must be in numeric range now"],
diff --git a/src/test/monolith_opcodes_tests.cpp b/src/test/monolith_opcodes_tests.cpp
index 8362fd541f..ce956f8b02 100644
--- a/src/test/monolith_opcodes_tests.cpp
+++ b/src/test/monolith_opcodes_tests.cpp
@@ -731,6 +731,13 @@ BOOST_AUTO_TEST_CASE(div_and_mod_opcode_tests) {
     CheckDivModError({{}}, ScriptError::INVALID_STACK_OPERATION);
 
     // CheckOps not valid numbers
+    CheckDivMod({0x01, 0x02, 0x03, 0x04, 0x05}, {0x01, 0x02, 0x03, 0x04, 0x05},
+                {0x01}, {});
+    CheckDivMod({0x01, 0x02, 0x03, 0x04, 0x05}, {0x01},
+                {0x01, 0x02, 0x03, 0x04, 0x05}, {});
+    CheckDivMod({0x01, 0x05}, {0x01, 0x02, 0x03, 0x04, 0x05},
+                {}, {0x01, 0x05});
+
     CheckDivModError({{0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09},
                       {0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09}},
                      ScriptError::INTEGER_OVERFLOW);
diff --git a/test/functional/logos_feature_int63.py b/test/functional/logos_feature_int63.py
new file mode 100755
index 0000000000..84b7025e34
--- /dev/null
+++ b/test/functional/logos_feature_int63.py
@@ -0,0 +1,225 @@
+#!/usr/bin/env python3
+# Copyright (c) 2021 The Logos developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test int63+1 bit integers."""
+
+import re
+
+from test_framework.txtools import pad_tx
+from test_framework.script import hash160
+from test_framework.blocktools import (
+    create_block,
+    create_coinbase,
+    make_conform_to_ctor,
+)
+from test_framework.messages import (
+    COutPoint,
+    CTransaction,
+    CTxIn,
+    CTxOut,
+)
+from test_framework.p2p import (
+    P2PDataStore,
+)
+from test_framework.script import (
+    CScript,
+    OP_1ADD,
+    OP_1SUB,
+    OP_NEGATE,
+    OP_ABS,
+    OP_NOT,
+    OP_0NOTEQUAL,
+    OP_ADD,
+    OP_SUB,
+    OP_DIV,
+    OP_MOD,
+    OP_LESSTHAN,
+    OP_GREATERTHAN,
+    OP_LESSTHANOREQUAL,
+    OP_GREATERTHANOREQUAL,
+    OP_HASH160,
+    OP_EQUAL,
+    OP_EQUALVERIFY,
+    OP_TRUE,
+)
+from test_framework.test_framework import BitcoinTestFramework
+from decimal import Decimal
+import random
+
+MAX_SCRIPT_INT = 0x7fff_ffff_ffff_ffff
+MIN_SCRIPT_INT = -0x7fff_ffff_ffff_ffff
+
+
+class Int63Test(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.setup_clean_chain = True
+        self.extra_args = [["-whitelist=noban@127.0.0.1"]]
+
+    def run_test(self):
+        node = self.nodes[0]
+        node.add_p2p_connection(P2PDataStore())
+        # OP_TRUE in P2SH
+        address = node.decodescript('51')['p2sh']
+        num_mature_coins = 30
+        node.generatetoaddress(num_mature_coins, address)
+        node.generatetoaddress(100, address)
+
+        value = 50 * 100_000_000
+        p2sh_script = CScript([OP_HASH160, bytes(20), OP_EQUAL])
+
+        def make_tx(coin_height):
+            assert coin_height <= num_mature_coins
+            block_hash = node.getblockhash(coin_height)
+            coin = int(node.getblock(block_hash)['tx'][0], 16)
+            tx = CTransaction()
+            tx.vin.append(
+                CTxIn(COutPoint(coin, 0), CScript([b'\x51'])))
+            return tx
+
+        def make_block():
+            parent_block_header = node.getblockheader(node.getbestblockhash())
+            coinbase = create_coinbase(parent_block_header['height'] + 1)
+            coinbase.vout[0].scriptPubKey = p2sh_script
+            coinbase.calc_sha256()
+            block = create_block(
+                int(parent_block_header['hash'], 16), coinbase, parent_block_header['time'] + 1)
+            return block
+
+        interesting_numbers = [
+            0,
+            1,
+            -1,
+            2,
+            -2,
+            4,
+            -4,
+            10,
+            -10,
+            127,
+            -127,
+            256,
+            -256,
+            0x7fffffff,
+            -0x7fffffff,
+            0x100000000,
+            -0x100000000,
+            0x7fffffffff,
+            -0x7fffffffff,
+            0x10000000000,
+            -0x10000000000,
+            0x7fffffffffffff,
+            -0x7fffffffffffff,
+            0x100000000000000,
+            -0x100000000000000,
+            0x7fffffffffffffff,
+            -0x7fffffffffffffff,
+            0x10000000000000000,
+            -0x10000000000000000,
+        ]
+
+        # make integer scripts
+        valid_scripts = []
+        invalid_scripts = []
+        def make_script(a, b = None, *, result, opcode):
+            if (MIN_SCRIPT_INT <= a <= MAX_SCRIPT_INT and
+                    (b is None or MIN_SCRIPT_INT <= b <= MAX_SCRIPT_INT) and
+                    (MIN_SCRIPT_INT <= result <= MAX_SCRIPT_INT)):
+                if b is None:
+                    valid_scripts.append(CScript([a, opcode, result, OP_EQUALVERIFY, OP_TRUE]))
+                else:
+                    valid_scripts.append(CScript([a, b, opcode, result, OP_EQUALVERIFY, OP_TRUE]))
+            else:
+                if b is None:
+                    invalid_scripts.append(CScript([a, opcode, OP_TRUE]))
+                else:
+                    invalid_scripts.append(CScript([a, b, opcode, OP_TRUE]))
+        for a in interesting_numbers:
+            make_script(a, result=a + 1, opcode=OP_1ADD)
+            make_script(a, result=a - 1, opcode=OP_1SUB)
+            make_script(a, result=-a, opcode=OP_NEGATE)
+            make_script(a, result=abs(a), opcode=OP_ABS)
+            make_script(a, result=not a, opcode=OP_NOT)
+            make_script(a, result=a != 0, opcode=OP_0NOTEQUAL)
+            for b in interesting_numbers:
+                make_script(a, b, result=a + b, opcode=OP_ADD)
+                make_script(a, b, result=a - b, opcode=OP_SUB)
+                if b != 0:
+                    # Note: We have to use Decimal here, as Python's integers behave differently
+                    # for division and modulo for negative numbers.
+                    make_script(a, b, result=int(Decimal(a) // Decimal(b)), opcode=OP_DIV)
+                    make_script(a, b, result=int(Decimal(a) % Decimal(b)), opcode=OP_MOD)
+                else:
+                    invalid_scripts.append(CScript([a, b, OP_DIV, OP_TRUE]))
+                    invalid_scripts.append(CScript([a, b, OP_MOD, OP_TRUE]))
+                make_script(a, b, result=a < b, opcode=OP_LESSTHAN)
+                make_script(a, b, result=a > b, opcode=OP_GREATERTHAN)
+                make_script(a, b, result=a <= b, opcode=OP_LESSTHANOREQUAL)
+                make_script(a, b, result=a >= b, opcode=OP_GREATERTHANOREQUAL)
+
+        txs = []
+        num_txs = 10
+        scripts_per_tx = len(valid_scripts) // num_txs
+        for i in range(1, num_txs + 1):
+            fund_tx = make_tx(i * 2)
+            spend_tx = make_tx(i * 2 + 1)
+            scripts = valid_scripts[(i - 1) * scripts_per_tx:][:scripts_per_tx]
+            for script in scripts:
+                fund_tx.vout.append(CTxOut(value // len(scripts),
+                    CScript([OP_HASH160, hash160(script), OP_EQUAL])))
+            fund_tx.rehash()
+            for i, script in enumerate(scripts):
+                spend_tx.vin.append(CTxIn(COutPoint(fund_tx.sha256, i), CScript([script])))
+            spend_tx.vout.append(CTxOut(value // len(scripts), p2sh_script))
+            txs.append(fund_tx)
+            txs.append(spend_tx)
+        block = make_block()
+        block.vtx.extend(txs)
+        make_conform_to_ctor(block)
+        block.hashMerkleRoot = block.calc_merkle_root()
+        block.solve()
+        node.p2p.send_blocks_and_test([block], node)
+        
+        fund_txs = []
+        invalid_spend_txs = []
+        num_txs = 5
+        scripts_per_tx = len(invalid_scripts) // num_txs
+        for i in range(1, num_txs + 1):
+            fund_tx = make_tx(21 + i)
+            scripts = invalid_scripts[(i - 1) * scripts_per_tx:][:scripts_per_tx]
+            for script in scripts:
+                fund_tx.vout.append(CTxOut(value // len(scripts),
+                    CScript([OP_HASH160, hash160(script), OP_EQUAL])))
+            fund_tx.rehash()
+            fund_txs.append(fund_tx)
+
+            for vout, script in enumerate(scripts):
+                spend_tx = CTransaction()
+                spend_tx.vin.append(CTxIn(COutPoint(fund_tx.sha256, vout), CScript([script])))
+                spend_tx.vout.append(CTxOut(value // len(scripts), p2sh_script))
+                pad_tx(spend_tx)
+                invalid_spend_txs.append(spend_tx)
+
+        block = make_block()
+        block.vtx.extend(fund_txs)
+        make_conform_to_ctor(block)
+        block.hashMerkleRoot = block.calc_merkle_root()
+        block.solve()
+        node.p2p.send_blocks_and_test([block], node)
+
+        invalid_block = make_block()
+        invalid_block.vtx.append(None)
+        for invalid_spend_tx in random.sample(invalid_spend_txs, 100):
+            invalid_block.vtx[1] = invalid_spend_tx
+            make_conform_to_ctor(invalid_block)
+            invalid_block.hashMerkleRoot = invalid_block.calc_merkle_root()
+            invalid_block.solve()
+            node.p2p.send_blocks_and_test(
+                [invalid_block], node,
+                success=False,
+                reject_reason='state=blk-bad-inputs, parallel script check failed')
+
+
+if __name__ == '__main__':
+    Int63Test().main()
diff --git a/test/functional/test_runner.py b/test/functional/test_runner.py
index 96d212dc41..1ac5e7727a 100755
--- a/test/functional/test_runner.py
+++ b/test/functional/test_runner.py
@@ -867,7 +867,7 @@ def check_script_prefixes(all_scripts):
     LEEWAY = 0
 
     good_prefixes_re = re.compile(
-        "(abc_)?(example|feature|interface|mempool|mining|p2p|rpc|wallet|tool|chronik)_"
+        "(abc_|logos_)?(example|feature|interface|mempool|mining|p2p|rpc|wallet|tool|chronik)_"
     )
     bad_script_names = [
         script for script in all_scripts if good_prefixes_re.match(script) is None
